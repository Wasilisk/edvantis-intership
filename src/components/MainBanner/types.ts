export type MainBannerStylesProps = {
    backgroundImage: string
}