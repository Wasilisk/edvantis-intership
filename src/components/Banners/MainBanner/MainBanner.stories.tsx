import {ComponentMeta, Story} from "@storybook/react";
import React from "react";
import {BannerType} from "../../../models";
import MainBanner from "./index";

const mockMainBanner: BannerType = {
    title: {
        highlight: "Furniture",
        after_highlight: "for Home & Office"
    },
    description: "500+ models of cushioned furniture, lamps and accessories.",
    image: {
        url: "https://res2.weblium.site/res/5e2f1da2c947890021c3f194/5e34250c4982e30021320262_optimized_1920"
    },
    link: {
        name: "Best-selling products",
        to: "/to"
    }
}

export default {
    title: 'Application/Banners/MainBanner',
    component: MainBanner,
} as ComponentMeta<typeof MainBanner>;

const Template: Story<BannerType> = (props) => <MainBanner {...props}/>;

export const Default = Template.bind({});
Default.args = {
    ...mockMainBanner
};
