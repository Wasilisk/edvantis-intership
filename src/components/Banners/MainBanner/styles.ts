import styled from "styled-components";
import {MainBannerStylesProps} from "./types";
import {LinkButtonStyles} from "../../common/LinkButton/styles";

export const MainBannerStyles = styled.div<MainBannerStylesProps>`
  height: calc(100vh - 102px);;
  max-width: 1200px;
  width: 100%;
  margin: 0 auto;
  padding: 0 15px;
  box-sizing: border-box;
  display: flex;
  align-items: center;
  
  .banner-info {
    max-width: 770px;

    h1 {
      font-size: 84px;
      color: #333333;
      font-family: Poppins,sans-serif;
      font-weight: 700;
      letter-spacing: 0;
      line-height: 1.1;
      margin: 0;
      overflow-wrap: anywhere;
    }
    p {
      font-size: 18px;
      overflow-wrap: anywhere;
      margin: 50px 0;
    }
  }
  
  .banner-image {
    height: 100vh;
    position: absolute;
    background-image: ${({backgroundImage}) => `url(${backgroundImage})`};
    background-size: cover;
    background-repeat: no-repeat;
    background-position: 0 70%;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    z-index: -1;
    overflow-x: hidden;
  }

  @media (max-width: 768px) {
    height: calc(100vh - 72px);
    .banner-info {
      h1 {
        font-size: 70px;
      }
    }
    .banner-image {
      height: 100vh;
    }
  }
  @media (max-width: 425px) {
    .banner-info {
      h1 {
        font-size: 58px;
      }
    }
    
    ${LinkButtonStyles} {
      width: auto;
    }
  }
`;