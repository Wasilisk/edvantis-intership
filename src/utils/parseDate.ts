const month = ["January","February","March","April","May","June","July","August","September","October","November","December"];

export const parseDate = (timestamp: number) => {
    const date = new Date(timestamp);

    return `${month[date.getMonth()]} ${('0' + date.getDate()).slice(-2)}, ${date.getFullYear()}`;
}