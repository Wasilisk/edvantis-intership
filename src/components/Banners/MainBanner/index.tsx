import React from 'react';
import {MainBannerStyles} from './styles';
import {useAppSelector} from "../../../hooks";
import {selectMainBanner} from "../../../store/slices/HomePage/selectors";
import LinkButton from "../../common/LinkButton";

const MainBanner = () => {
    const mainBannerData = useAppSelector(selectMainBanner);

    return (
        mainBannerData && <MainBannerStyles backgroundImage={mainBannerData.image.url}>
            <div className="banner-info">
                <h1>
                    <span className="highlight">{mainBannerData.title.highlight}&nbsp;</span>
                    {mainBannerData.title.plain_text}
                </h1>
                <p>{mainBannerData.description}</p>
                <LinkButton
                    data-testid="link-button"
                    to={mainBannerData.link.to}
                    variant="filled"
                    size="large"
                    label={mainBannerData.link.name}
                    hasArrow
                    fullWidth={false}
                />
            </div>
            <div className="banner-image" data-testid="banner"/>
        </MainBannerStyles>
    );
};

export default MainBanner;
