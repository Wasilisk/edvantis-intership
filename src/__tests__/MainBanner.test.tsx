import MainBanner from "../components/Banners/MainBanner";
import {render} from "./utils/store-wrapper";
import {mockMainBanner} from "./__mock__";
import {unstable_HistoryRouter as HistoryRouter} from "react-router-dom";
import {createMemoryHistory} from "history";
import {fireEvent} from "@testing-library/react";

describe('MainBanner component', () => {
    it('should render component', () => {
        const history = createMemoryHistory({ initialEntries: ['/'] });
        const {getByText, asFragment, getByTestId} = render(
            <HistoryRouter history={history}>
                <MainBanner/>
            </HistoryRouter>, {
            initialState: {
                home_page: {
                    main_banner: mockMainBanner,
                    isLoading: false
                }
            }
        });
        expect(getByTestId('banner')).toHaveStyle(`background-image: url(${mockMainBanner.image.url})`);

        const highlightText = getByText(mockMainBanner.title.highlight!);
        expect(highlightText).toBeDefined();

        const plainText = getByText(mockMainBanner.title.after_highlight!);
        expect(plainText).toBeDefined();

        const linkButton = getByTestId('link-button');
        expect(linkButton).toHaveTextContent(mockMainBanner.link.name);
        fireEvent.click(linkButton);
        expect(history.location.pathname).toBe(mockMainBanner.link.to);

        expect(asFragment()).toMatchSnapshot();
    })
    it('should be empty when data is loading', () => {
        const {container} = render(<MainBanner/>, {
            initialState: {
                home_page: {
                    main_banner: null,
                    isLoading: true
                }
            }
        });

        expect(container.childElementCount).toEqual(0);
    })
})