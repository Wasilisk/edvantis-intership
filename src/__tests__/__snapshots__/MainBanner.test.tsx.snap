// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainBanner component should render component 1`] = `
<DocumentFragment>
  <div
    class="sc-gsnTZi jaaNkd"
  >
    <div
      class="banner-info"
    >
      <h1>
        <span
          class="highlight"
        >
          Highlight text 
        </span>
        Simple text
      </h1>
      <p>
        Main banner description
      </p>
      <a
        class="sc-bczRLJ gLqqtr"
        href="/home/link"
      >
        <button
          class="sc-dkzDqf hTVmxt"
          data-testid="link-button"
        >
          Link name
          <svg
            height="100%"
            viewBox="0 0 36 36"
            width="100%"
            x="0px"
            xmlns="http://www.w3.org/2000/svg"
            y="0px"
          >
            <path
              class="cls-1"
              d="M34,18l-5-4v3H26v2h3v3ZM24,19H20V17h4v2Zm-6,0H14V17h4v2Zm-6,0H8V17h4v2ZM4,16a2,2,0,1,1-2,2A2,2,0,0,1,4,16Z"
              data-name="28"
              fill="currentColor"
            />
          </svg>
        </button>
      </a>
    </div>
    <div
      class="banner-image"
      data-testid="banner"
    />
  </div>
</DocumentFragment>
`;
